### ISSUE TITLE
[CUDA] Reduction extremely slow

### ISSUE THREAD

**Issue by silvasean** [iree-org/iree#13285]
### What happened?

The reduction below takes ~120ms on my A100 to reduce a tensor of size 833x833xf32 which seems off by over 1000x (an 800x800x800 matmul is measured in microseconds on this chip) . I tried making the size 832 to see if alignment was the problem but that did not significantly change the runtime, so I think something more fundamental is going on.

Repro with:

```
iree-compile --iree-hal-target-backends=cuda --iree-input-type=mhlo --iree-hal-cuda-llvm-target-arch=sm_80 trace_inner.mlir -o trace_inner.vmfb
iree-benchmark-module --function=main --module=trace_inner.vmfb --device=cuda --input=833xi32=0 --input=833x833xf32=0 --input=f32=0
```

```
#map = affine_map<(d0, d1) -> (d0)>
#map1 = affine_map<(d0, d1) -> (d1)>
#map2 = affine_map<(d0, d1) -> (d0, d1)>
#map3 = affine_map<(d0, d1) -> ()>
module {
  func.func @main(%arg0: tensor<833xi32>, %arg1: tensor<833x833xf32>, %arg2: tensor<f32>) -> tensor<f32> {
    %cst = arith.constant 5.66893432E-4 : f32
    %0 = tensor.empty() : tensor<f32>
    %1 = linalg.fill ins(%cst : f32) outs(%0 : tensor<f32>) -> tensor<f32>
    %2 = linalg.generic {indexing_maps = [#map, #map1, #map2, #map3, #map3], iterator_types = ["reduction", "reduction"]} ins(%arg0, %arg0, %arg1, %arg2 : tensor<833xi32>, tensor<833xi32>, tensor<833x833xf32>, tensor<f32>) outs(%1 : tensor<f32>) {
    ^bb0(%in: i32, %in_0: i32, %in_1: f32, %in_2: f32, %out: f32):
      %3 = arith.divf %in_1, %in_2 : f32
      %4 = arith.cmpi eq, %in, %in_0 : i32
      %5 = arith.select %4, %3, %cst : f32
      %6 = arith.addf %out, %5 : f32
      linalg.yield %6 : f32
    } -> tensor<f32>
    return %2 : tensor<f32>
  }
}
```

### Steps to reproduce your issue

See above

### What component(s) does this issue relate to?

Compiler

### Version information

iree @ ab37989652aed11f7f46498c09b9ac515c83eaa3

**Comment by allieculp** [comment]
@mattwalsh @julianwa to help triage.

**Comment by julianwa** [comment]
@okkwon can you root cause and fix if straightforward and send back to triage otherwise?

**Comment by allieculp** [comment]
@okkwon Let us know when you have had a chance to review this.

**Comment by okkwon** [comment]
I will take a look in this week.

**Comment by okkwon** [comment]
@julianwa @allieculp can one of you set the priority? I thought it was P2 and that's why I put it in my "look-at-later" queue.

**Comment by okkwon** [comment]
The pattern looks similar to the other case, where EW + reduction prevented the codegen from choosing the tensorcore. https://github.com/openxla/iree/pull/13308 may fix the issue too. 

Here is the other case: https://github.com/openxla/iree/issues/13211

**Comment by okkwon** [comment]
@silvasean can you provide the input mlir before FusionOfTensor pass? #13308 may have fixed the issue. You can attach the original model too.

**Comment by okkwon** [comment]
The reduction is slow because it is fully serialized without any tiling.

https://gist.github.com/okkwon/acd553970efcc4e94101a716aa87ffe4

**Comment by okkwon** [comment]
More analysis:
1. Making the tensor aligned (832x832) does not help generate parallel code.
2. Removing EW ops and putting a single reduction (833x833) does not help generate parallel code.
3. Single op with aligned reduction does generate parallel code.

So, we have a mixed situation. I would approach this way.
1. support unaligned reduction case.
2. change the fusion to not fuse EW operation producer with reduction consumer. #13308 may help. Or, make the reduction codegen path more robust to support this case.

@julianwa please prioritize the issue and assign it to someone. An unaligned reduction support may need a substantial code change.

**Comment by allieculp** [comment]
@julianwa @mattwalsh Marking as P1, needs scheduling. We will need a resource here.

**Comment by okkwon** [comment]
I heard that unaligned reductions are supposed to be supported by transform-dialect. Is it true @nicolasvasilache?

**Comment by julianwa** [comment]
> @julianwa @mattwalsh Marking as P1, needs scheduling. We will need a resource here.

Priority sounds right. We can expect upwards of 30-50% e2e slow down on real world models that hit this case. This should be considered for next sprint.

**Comment by allieculp** [comment]
@mattwalsh Adding this to our Github roadmap.
@okkwon FYI Nicolas is OOO this week.

**Comment by nicolasvasilache** [comment]
The reduction strategies would kick in for unaligned cases if the linalg op had 1, 2 or 3 parallel iterators + 1 reduction iterator.
Here the linalg op only has 2 reduction iterators and 0 parallel iterators.

Performing a split-K of size 1 on the first iterator will turn this into `parallel, reduction` followed by `reduction`, on which the strategies could then apply.

Note that this would be a pre-processing transformation and I am not aware of a good place to add such pre-processings in IREE atm.

**Comment by silvasean** [comment]
I'm hitting a similar issue in another model with size 2048x2048: https://gist.github.com/silvasean/cf1061249f8cb008e91b511da76e7bf5
(taking 11ms for reducing an 2048x2048 matrix to scalar)

**Comment by allieculp** [comment]
@ThomasRaoux can you take a look at this?
@mattwalsh for vis

**Comment by silvasean** [comment]
Here is another one: 1d->0d reduction
https://gist.github.com/silvasean/6ea40a150c8c87c490e2b7a9429d89e3

**Comment by nicolasvasilache** [comment]
FYI, a slight variation of the solution proposed here should help: https://github.com/openxla/iree/issues/13115#issuecomment-1543844223

I can take a stab at a first cut tomorrow.

**Comment by allieculp** [comment]
@nicolasvasilache Any update on this one?

**Comment by nicolasvasilache** [comment]
I investigated this deeper and what I proposed in #13115 does not help for the motivating example that Sean posted in this bug report.

The reason is that the existing split-K transformations supports a single reduction iterator.
It is unclear how to extend the existing implementation to allow multiple reductions.

Additionally, we cannot just reshape this fused mixed broadcast / reduction example into a 1-D -> 0-D reduction AFAICS.

A potential path forward could be to implement a staged reduction rewrite that would rewrite e.g. a 3-D -> 0-D reduction (strawman):
```
[red, red, red] -> [par, par, red] + [par, red] + [red]
```
by introducing a 2-D and a 1-D intermediate tensor.
This is however something to investigate separately from #13115.

Lastly, I see this comment from @okkwon:
```
3. Single op with aligned reduction does generate parallel code.
```
Could you elaborate what "single op" means here ? 
Does it mean that 
```
    %2 = linalg.generic {indexing_maps = [#map, #map1, #map2, #map3, #map3], iterator_types = ["reduction", "reduction"]} ins(%arg0, %arg0, %arg1, %arg2 : tensor<832xi32>, tensor<832xi32>, tensor<832x832xf32>, tensor<f32>) outs(%1 : tensor<f32>) {
    ^bb0(%in: i32, %in_0: i32, %in_1: f32, %in_2: f32, %out: f32):
      %3 = arith.divf %in_1, %in_2 : f32
      %4 = arith.cmpi eq, %in, %in_0 : i32
      %5 = arith.select %4, %3, %cst : f32
      %6 = arith.addf %out, %5 : f32
      linalg.yield %6 : f32
    } -> tensor<f32>
```
in isolation without the fill generates parallel code?
If so, does it generate both block and thread parallelism?
A pointer to where that happens in IREE would be useful.

**Comment by nicolasvasilache** [comment]
Separately, looking deeper into the computation, I see this computes:
```
o += sum_i_j (%arg0[i] == %arg0[j]) ? (%arg1[i][j] / %arg2) : %cst 
```

This could also be achieved by precomputing:
```
tmp[i, j] =  (%arg0[i] == %arg0[j]) ? (%arg1[i][j] / %arg2) : %cst 
o += sum_i_j tmp[i, j]
```

The first operation is a 2-D elementwise and the second can be reshaped to a 1-D reduction on which flavors of split-K apply.

So this makes me wonder how we obtained this op in the first place @silvasean and whether it is possible we hit a case where `linalg.generic` can represent such a computation with aggressive fusion but this may not be desired?

**Comment by allieculp** [comment]
@MaheshRavishankar for vis

**Comment by MaheshRavishankar** [comment]
@nicolasvasilache looking at the original comment, this op seems to be how it is lowered into Linalg from frontend dialects. @silvasean do you have something from stablehlo?

**Comment by silvasean** [comment]
Here is something similar to the example from the OP in StableHLO: https://gist.github.com/silvasean/12afdda5dfd8809a99600a1a25350687

It comes from this JAX program:
```
import jax.numpy as jnp
import jax

@jax.jit
def trace(a):
    return jnp.trace(a)

trace(jnp.ones([2048, 2048]))
```

It comes from JAX lowering matrix trace as masking off the non-diagonal of the matrix (comparing iota(dim=0) == iota(dim=1)) and then reducing the whole matrix. I know we can represent this better in linalg with linalg.index, but it happens to hit this reduction issue, which is present in other workloads, so it seems worth addressing independently of the better representation. For example in LLM (or really any ML model) training, frequently one saves off summaries of tensors (like the mean of the whole tensor or other statistics) which is important for modelers to track the training progress.

**Comment by MaheshRavishankar** [comment]
> Here is something similar to the example from the OP in StableHLO: [gist.github.com/silvasean/12afdda5dfd8809a99600a1a25350687](https://gist.github.com/silvasean/12afdda5dfd8809a99600a1a25350687)
> 
> It comes from this JAX program:
> 
> ```
> import jax.numpy as jnp
> import jax
> 
> @jax.jit
> def trace(a):
>     return jnp.trace(a)
> 
> trace(jnp.ones([2048, 2048]))
> ```
> 
> It comes from JAX lowering matrix trace as masking off the non-diagonal of the matrix (comparing iota(dim=0) == iota(dim=1)) and then reducing the whole matrix. I know we can represent this better in linalg with linalg.index, but it happens to hit this reduction issue, which is present in other workloads, so it seems worth addressing independently of the better representation. For example in LLM (or really any ML model) training, frequently one saves off summaries of tensors (like the mean of the whole tensor or other statistics) which is important for modelers to track the training progress.

(not saying we shouldnt address the issue described here, but more for my understanding). Wont these reductions have a batch dimension?

**Comment by silvasean** [comment]
No, they don't have a batch dimension in all cases. Both of the examples I gave in this thread are from real workloads.

**Comment by silvasean** [comment]
In the example tensor summary example it is easy to see that there is no batch dimension, since the point is to reduce the whole tensor to a single value for statistics reporting. I don't know for sure why the FermiNet one does not have a batch dimension, but I suspect it is due to being a global value computed somewhere in the KFAC optimizer.

**Comment by allieculp** [comment]
@nicolasvasilache Can you provide an update here on Monday?

**Comment by nicolasvasilache** [comment]
We either need a new transform as I described, alternatively, we could make the lowering from xxHLO more friendly by lowering to the form I laid out. 

This will need a new owner, as this is not covered by the same transform that I was thinking about originally.

**Comment by allieculp** [comment]
@mattwalsh See above notes, we will need a new owner for this.

**Comment by MaheshRavishankar** [comment]
Ok, had a conversation with @silvasean about this. So the initial bug report is a bit confusing cause it seems like this is not the "input" from stablehlo lowering, but instead is actually the dispatch that is slow and was extracted out.... In that case, I can see how this `linalg.generic` was a result of aggressive fusion.  Ill take this one over.

**Comment by MaheshRavishankar** [comment]
With #13730 this addresses the comment https://github.com/openxla/iree/issues/13285#issuecomment-1549166170 . Now the issues comes down to single large reduction dimension that makes it same as #13115 

I am moving this back to Nicolas since the blocker to make this addressable through #13115 is now removed.

**Comment by allieculp** [comment]
@silvasean Downgrading this from P0 to P1 as well - please let us know if that changes.

**Comment by nicolasvasilache** [comment]
> With #13730 this addresses the comment [#13285 (comment)](https://github.com/openxla/iree/issues/13285#issuecomment-1549166170) . Now the issues comes down to single large reduction dimension that makes it same as #13115
> 
> I am moving this back to Nicolas since the blocker to make this addressable through #13115 is now removed.

Ok, the intuition actually held that we can reduce to #13115 by avoiding too much fusion + reshape.
Thanks for investigating!

Can we actually close this bug completely (@silvasean ) ?

**Comment by silvasean** [comment]
Sure, we can close it. If we do that then I would prefer if we incorporated this use case (or some unfused modification of it, which I think Mahesh has locally) as an acceptance criterion for #13115 being closed though just to make sure everything connects.
